src/lib/Form/Type/FieldType/AbstractRelationFieldType.php

```diff
@@ -0,0 +1,134 @@
+            $contentInfo = $versionInfo->getContentInfo();
+            $contentType = $contentInfo->getContentType();
+            $languageCodes = $this->getAvailableLanguageCodes($contentInfo, $versionInfo->getLanguages());
+        } catch (UnauthorizedException $e) {
+            $unauthorized = true;
+        }
+
+        return [
+            'contentInfo' => $contentInfo,
+            'contentType' => $contentType,
+            'contentId' => $contentId,
+            'languageCodes' => $languageCodes,
+            'unauthorized' => $unauthorized,
+        ];
+    }
+
+    /**
+     * @param array<\Ibexa\Contracts\Core\Repository\Values\Content\Language> $languages
+     *
+     * @return array<string>
+     */
+    public function getAvailableLanguageCodes(ContentInfo $contentInfo, array $languages): array
+    {
+        $languageCodes = [];
+        foreach ($languages as $language) {
+            $languageCodes[] = $language->getLanguageCode();
+        }
+
+        $lookupLimitationResult = $this->getLookupLimitationResult($contentInfo, $languageCodes);
+        if (
+            empty($lookupLimitationResult->lookupPolicyLimitations)
+            && empty($lookupLimitationResult->roleLimitations)
+        ) {
+            return $languageCodes;
+        }
+
+        $limitationLanguageCodes = $this->getLimitationLanguageCodes($lookupLimitationResult);
+
+        return array_filter(
+            $languageCodes,
+            static function (string $languageCode) use ($limitationLanguageCodes): bool {
+                return in_array($languageCode, $limitationLanguageCodes, true);
+            }
+        );
+    }
+
+    private function getLookupLimitationResult(ContentInfo $contentInfo, array $languageCodes): LookupLimitationResult
+    {
+        return $this->permissionResolver->lookupLimitations(
+            'content',
+            'edit',
+            $contentInfo,
+            [
+                (new Target\Builder\VersionBuilder())->translateToAnyLanguageOf($languageCodes)->build(),
+                $contentInfo->getMainLocation(),
+            ],
+            [Limitation::LANGUAGE]
+        );
+    }
+
+    /**
+     * @return array<string>
+     */
+    private function getLimitationLanguageCodes(LookupLimitationResult $lookupLimitations): array
+    {
+        $limitationLanguageCodes = [];
+        foreach ($lookupLimitations->roleLimitations as $roleLimitation) {
+            $limitationLanguageCodes[] = $roleLimitation->limitationValues;
+        }
+
+        foreach ($lookupLimitations->lookupPolicyLimitations as $lookupPolicyLimitation) {
+            foreach ($lookupPolicyLimitation->limitations as $limitation) {
+                $limitationLanguageCodes[] = $limitation->limitationValues;
+            }
+        }
+
+        return !empty($limitationLanguageCodes)
+            ? array_unique(array_merge(...$limitationLanguageCodes))
+            : $limitationLanguageCodes;
+    }
```

Comment:
These will always be codes for languages? :thinking: 

Regardless of the above, I would suggest using a "hashmap" to work unique language codes without the need of creating potentially large arrays + merge (large because of multiple roles/limitations).

Basically, add every language code as
```php
$limitationLanguageCodes[$languageCode] = true;
```
and then simply
```php
return array_keys($limitationLanguageCodes);
```
at the end.

Reply:
I checked Language limitation in doc (https://doc.ibexa.co/en/latest/permissions/limitation_reference/#language-limitation) before I started adding this code and limitation value will always be a code for language. 

Fixed: https://github.com/ibexa/content-forms/pull/51/commits/d23c440332e73cd8b7969f0911f8333d8d02c710